source/common/upstream/health_checker_impl.cc

```diff
@@ -281,15 +284,52 @@ void HttpHealthCheckerImpl::HttpActiveHealthCheckSession::onInterval() {
 
 void HttpHealthCheckerImpl::HttpActiveHealthCheckSession::onResetStream(Http::StreamResetReason,
                                                                         absl::string_view) {
+  request_in_flight_ = false;
+  ENVOY_CONN_LOG(debug, "connection/stream error health_flags={}", *client_,
+                 HostUtility::healthFlagsToString(*host_));
   if (expect_reset_) {
     return;
   }
 
-  ENVOY_CONN_LOG(debug, "connection/stream error health_flags={}", *client_,
-                 HostUtility::healthFlagsToString(*host_));
+  if (client_ && !reuse_connection_) {
+    client_->close();
+  }
+
   handleFailure(envoy::data::core::v3::NETWORK);
 }
 
+void HttpHealthCheckerImpl::HttpActiveHealthCheckSession::onGoAway(
+    Http::GoAwayErrorCode error_code) {
+  ENVOY_CONN_LOG(debug, "connection going away goaway_code={}, health_flags={}", *client_,
+                 error_code, HostUtility::healthFlagsToString(*host_));
+
+  // Runtime guard around graceful handling of NO_ERROR GOAWAY handling. The old behavior is to
+  // ignore GOAWAY completely.
+  if (!parent_.runtime_.snapshot().runtimeFeatureEnabled(
+          "envoy.reloadable_features.health_check.graceful_goaway_handling")) {
+    return;
+  }
+
+  if (request_in_flight_ && error_code == Http::GoAwayErrorCode::NoError) {
+    // The server is starting a graceful shutdown. Allow the in flight request
+    // to finish without treating this as a health check error, and then
+    // reconnect.
+    reuse_connection_ = false;
+    return;
+  }
+
+  if (request_in_flight_) {
+    // Record this as a failed health check.
+    handleFailure(envoy::data::core::v3::NETWORK);
+  }
+
+  parent_.stats_.goaway_error_.inc();
```

Comment:
@mattklein123 this line has to be after `handleFailure()` or my test will flake. If you dig into `handleFailure()`, it increments the general health check failure stat, and then enables the interval timer. So order is:

1) health failure increment
2) health check timer enabled
3) goaway_error increment.

I need something that will reliably happen after the health check timer. 999/1000 times, waiting for client disconnect is good enough (because of the `client_->close()` on line 329) but the other 1 time the fake upstream was closing first, which would cause my test to advance too quickly and advance simulated time before the timer was enabled.

Reply:
OK really sorry for all the back and forth here, but this is pretty tricky and I'm trying to leave the code as clean as possible.

I think there are two options here that are both fine:
1) Swap the order of (1) and (2). I can't see any huge reason this would cause a problem and if it makes the tests cleaner that is fine with me.
2) Stop using simulated time. I think that would also work since you wouldn't have to actually advance and you could just wait?

I would be fine either way but I think this is simpler?

Reply:
No worries about the back and forth! I'm happy to do whatever it takes to get it right.

Both of those options should work, using real time might make the test a bit slower though. I couldn't tell in your last sentence which one you were saying is simpler?

Reply:
Either is fine with me. I think they are both simple and I'm not worried about test speed when using real time.